File: .vscode/linked_list/reverse_dll.py
class Node:
    def __init__(self,data):
        self.data=data
        self.next=None
        self.prev=None
def insert(head,d):
    n=Node(d)
    t=head
    if t==None:
        return n
    else:
        while(t.next!=None):
            t=t.next
        t.next=n
        n.prev=t
        return head
def reverse(head):     
     ptr=head
     p=None
     while (ptr!=None):
         n=ptr.next
         ptr.prev=ptr.next
         ptr.next=p
         p=ptr
         ptr=n
     return p
def display(head):
    if(head==None):
        print("Empty")
    while(head):
        print(head.data,end=" ")
        head=head.next
head=None
head=insert(head,10)
head=insert(head,30)
head=insert(head,50)
display(head)
print("\n")
head=reverse(head)
display(head)

File: linked_list/linked_list/reverse_dll.c
#include<stdio.h>
#include<stdlib.h>
typedef struct Node{
    int data;
    struct Node *next;
    struct Node *prev;
}Node;
Node *insert(Node *head,int d){
    Node *n=(Node *)calloc(1,sizeof(Node));
    n->data=d;
    Node *temp=head;
    if(temp==NULL){
        return n;
    }
    else{
        while(temp->next!=NULL){
        temp=temp->next;
        }
        temp->next=n;
        n->prev=temp;
        n->next=NULL;
        return head;
        // head->prev=NULL;
    }
}
    Node *reverse(Node *head){
        Node *ptr=head;
        Node *p=NULL;
        Node *n=head;
        // p->prev=NULL;
        while(ptr!=NULL){
           n=ptr->next;
           ptr->prev=ptr->next;
           ptr->next=p;
           p=ptr;
           ptr=n;
        }
        // ptr->next=p;
        return p;
    }
void display(Node *head){
    if (head==NULL){
        printf("Null");
    }
    else{
        while(head!=NULL){
            printf("%d->",head->data);
            head=head->next;
        }
    }
}
int main(){
    Node *head=NULL;
    head=insert(head,20);
    head=insert(head,30);
    head=insert(head,40);
    // display(head);
    head=reverse(head);
    display(head);
}
